test/test_newtons_method.jl

```diff
@@ -0,0 +1,68 @@
+using ClimaTimeSteppers, LinearAlgebra, Random, Test
+
+function linear_equation(FT, n)
+    rng = MersenneTwister(1)
+    A = rand(rng, FT, n, n)
+    b = rand(rng, FT, n)
+    f!(f, x) = f .= A * x .- b
+    j!(j, x) = j .= A
+    x_exact = A \ b
+    x_init = zeros(FT, n)
+    return (f!, j!, x_exact, x_init)
+end
+function nonlinear_equation(FT, n)
+    rng = MersenneTwister(1)
+    A = rand(rng, FT, n, n)
+    b = rand(rng, FT, n)
+    f!(f, x) = f .= A * sinh.(A * x .- b) .- b
+    j!(j, x) = j .= A * diagm(cosh.(A * x .- b)) * A
+    x_exact = A \ (asinh.(A \ b) .+ b)
+    x_init = zeros(FT, n)
+    return (f!, j!, x_exact, x_init)
+end
+
+@testset "Newton's Method" begin
+    # For each test case, step_adjustment is set to the value that minimizes the
+    # number of Newton iterations and the normed error of the last Newton
+    # iteration (for the specified value of rtol). In general, it is large when
+    # the corresponding function f is very smooth or has a large roundoff error.
+    for (is_linear, FT, n, step_adjustment) in (
+        (true, Float32, 1, 100000),
+        (true, Float32, 10, 1000000),
+        (true, Float64, 1, 1000000000),
+        (true, Float64, 10, 10000000000),
+        (false, Float32, 1, 1),
+        (false, Float32, 10, 1),
+        (false, Float64, 1, 0.1),
+        (false, Float64, 10, 10),
+    )
```

Comment:
For the linear equation, can you get machine error after 1 newton iteration?

Reply:
What is the step_adjustment?

Reply:
Yes, the linear equation can get down to `eps(FT)`, but the nonlinear equation only got down to `100eps(FT)`, which is why I used `100eps(FT)` for both. The linear equation is solved in 1 newton iteration, but the nonlinear one needs more, so I used 10. Step adjustment is explained in the documentation: https://clima.github.io/ClimaTimeSteppers.jl/previews/PR79/newtons_method/#ClimaTimeSteppers.ForwardDiffStepSize1

Reply:
That looks good to me! I think in ClimaMachine, we also tried tendency dependent step_adjustment, for example step_adjustment = ||tendency||

Reply:
Oh, interesting. All the papers on JFNK that I looked at only based the step size on `Y` and `ΔY`, and I'm just using `step_adjustment` as a small fudge factor to adjust those step sizes. The step size that I set as a default seems to work really well without any adjustment for non-linear problems; I've been using it to run IMEX EDMF with `step_adjustment = 1`.

Reply:
If you have a formula for a tendency-based step size, I could add that as well.

Reply:
I cannot find references, but we have explored this idea in ClimaMachine https://github.com/CliMA/ClimateMachine.jl/blob/2f84c00ecec8cc6b1783306327397c0879f643c6/src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl#L51



```
"""
Approximates the action of the Jacobian of a nonlinear
form on a vector `ΔQ` using the difference quotient:
      ∂F(Q)      F(Q + e ΔQ) - F(Q)
JΔQ = ---- ΔQ ≈ -------------------
       ∂Q                e
Compute  JΔQ with cached Q and F(Q), and the direction  dQ
"""
function (op::JacobianAction)(JΔQ, dQ, args...)
    rhs! = op.rhs!
    Q = op.Q
    Qdq = op.Qdq
    ϵ = op.ϵ
    Fq = op.Fq
    Fqdq = op.Fqdq

    FT = eltype(dQ)
    n = length(dQ)
    normdQ = norm(dQ, weighted_norm)

    β = √ϵ

    if normdQ > ϵ
        # for preconditioner reconstruction, it goes into this part
        # e depends only on the active freedoms in the preconditioner reconstruction
        factor = FT(1 / (n * normdQ))
        Qdq .= Q .* (abs.(dQ) .> 0)
        e = factor * β * norm(Qdq, 1, false) + β
    else
        factor = FT(1 / n)
        e = factor * β * norm(Q, 1, false) + β
    end



    Qdq .= Q .+ e .* dQ

    rhs!(Fqdq, Qdq, args...)

    JΔQ .= (Fqdq .- Fq) ./ e

end
```